lib/ProductOpener/Display.pm

```diff
@@ -8634,79 +8635,101 @@ the rounded value according to the Nutri-Score rules, and the corresponding poin
-		$html .= "<p>" . lang("nutriscore_proteins_maximum_fruits_points") . "</p>";
+				
+		$nutriscore_protein_info = lang("nutriscore_proteins_maximum_fruits_points");
 	}
 	else {
-		$html .= "<p>" . lang("nutriscore_proteins_negative_points_greater_or_equal_to_11") . "</p>";
-	}
-
-	$html .= "<p><strong>" . lang("nutriscore_score") . lang("sep"). ": " . ($nutriscore_data_ref->{score})
-	. "</strong> (" . $nutriscore_data_ref->{negative_points} . " - " . $nutriscore_data_ref->{positive_points} . ")<p>";
-
-	$html .= "<p><strong>" . lang("nutriscore_grade") . lang("sep"). ": " . uc($nutriscore_data_ref->{grade}) . "</strong></p>";
-
-	$html .= "</div>";
-
+		$nutriscore_protein_info = lang("nutriscore_proteins_negative_points_greater_or_equal_to_11");
+	}
+	
+	# Generate a data structure that we will pass to the template engine
+	
+	my $template_data_ref = {
+		
+		lang => \&lang,
+				
+		is_beverage => $nutriscore_data_ref->{is_beverage},
+		is_fat => $nutriscore_data_ref->{is_fat},
+		
+		nutriscore_protein_info => $nutriscore_protein_info,
+		
+		score => $nutriscore_data_ref->{score},
+		grade => uc($nutriscore_data_ref->{grade}),
+		positive_points => $nutriscore_data_ref->{positive_points},
+		negative_points => $nutriscore_data_ref->{negative_points},
+		
+		# Details of positive and negative points, filled dynamically below
+		# as the nutrients and thresholds are different for some products (beverages and fats)
+		points_groups => []
+	};
+	
+	my %points_groups = (
+		"positive" => ["proteins", "fiber", "fruits_vegetables_nuts_colza_walnut_olive_oils"],
+		"negative" => ["energy", "sugars", "saturated_fat", "sodium"],
+	);
+	
+	foreach my $type ("positive", "negative") {
+				
+		# Initiate a data structure for the points of the group
+		
+		my $points_group_ref = {
+			type => $type,
+			points => $nutriscore_data_ref->{$type . "_points"},
+			nutrients => [],
+		};
+		
+		# Add the nutrients for the group
+		foreach my $nutrient (@{$points_groups{$type}}) {
+			
+			my $nutrient_threshold_id = $nutrient;
+			
+		 	if ((defined $nutriscore_data_ref->{is_beverage}) and ($nutriscore_data_ref->{is_beverage})
+		 		and (defined $points_thresholds{$nutrient_threshold_id . "_beverages"})) {
+		 		$nutrient_threshold_id .= "_beverages";
+		 	}
+		 	if (($nutriscore_data_ref->{is_fat}) and ($nutrient eq "saturated_fat")) {
+		 		$nutrient = "saturated_fat_ratio";
+		 		$nutrient_threshold_id = "saturated_fat_ratio";
+		 	}
+			push @{$points_group_ref->{nutrients}}, {
+				id => $nutrient,
+				points => $nutriscore_data_ref->{$nutrient . "_points"},
+				maximum => scalar(@{$points_thresholds{$nutrient_threshold_id}}),
+				value => $nutriscore_data_ref->{$nutrient},
+				rounded => $nutriscore_data_ref->{$nutrient . "_value"},
+			};
+		 }
+		 
+		 push @{$template_data_ref->{points_groups}}, $points_group_ref;
+	 }
+	# Nutrition Score Calculation Template
+		my $config = {
```

Comment:
```suggestion
	my $config = {
```
Also, aren't these config we can use for every template? Can't we put them outside of the sub?